### PR TITLE
gtk: add Windows support

### LINKED ISSUE #10505: [package] gtk/4.4.0: shared option does not work
The gtk/4.4.0 package has only a .so library even with shared=False

### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **gtk/4.4.0**
  * Operating System+version: **Linux Ubuntu 20.04**
  * Compiler+version: **GCC 9**
  * Conan version: **conan 1.47.0**
  * Python version: **Python 3.8.10**

### Conan profile (output of `conan profile show default` or `conan profile show <profile>` if custom profile is in use)
```
Configuration for profile default:

[settings]
os=Linux
os_build=Linux
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=9
compiler.libcxx=libstdc++
build_type=Release
[options]
[conf]
[build_requires]
[env]
```

### Steps to reproduce (Include if Applicable)
First install gtk/4.4.0
```
conan install gtk/4.4.0@ --build=gtk --build=graphene -g txt -o gtk:shared=False # graphene is built to fix undefined reference to __atan2f_finite
```

Second, find the libpath of gtk from conanbuildinfo.txt, and inspect the path:
```
ls /home/zhaoquan/.conan/data/gtk/4.4.0/_/_/package/5c17355d7562eb3e21680750e9c180893d53bca6/lib
```

We can see that the lib directory has only .so but no .a, which means the libraries are all shared.
```
gtk-4.0  libgtk-4.so  libgtk-4.so.1  libgtk-4.so.1.400.0
```

### Logs (Include/Attach if Applicable)
N/A

### PR DESCRIPTION
**Pull request by taoyouh**
Specify library name and version:  **gtk/4.4.0**

Add Windows support for gtk recipe (gtk 4.2 or higher)

Fixes #10505 

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by ericLemanissier** [comment]
as a reference: the issues I faced when I tried: https://github.com/conan-io/conan-center-index/issues/4306

**Comment by taoyouh** [comment]
### About the issues mentioned in #4306

> 1. compilation of first file fails with "is the command line too long?"

With GTK 4, the gnome.mkenums triggers a similar failure, which is fixed by `short_paths = True`.

> 2. it fails later at link stage with`fatal error LNK1170: line in command file contains 131071 or more characters`
this is solved by using `CONAN_USE_ALWAYS_SHORT_PATHS=1` environment variable.

My local test doesn't have this problem.

> 3. duplicate definition of DllMain

Just like the issue said, I set `-o gdk-pixbuf:shared=True` and fixed it.

> 4. last failure happens during install

My tests with 4.3 and 4.4 don't have this problem.

### Some other problems I ran into
When I try to run the getting-started example of GTK, my app crashes at first. I debugged into it and found that cairo seems not initialized property, making a  segmentation fault. Adding `cairo:shared=True` fixed this.

### Current status
In my local build, I can launch a GTK window with all options default, compiler=VS2019, build_type=Debug/Release, version=4.4.*

**Comment by taoyouh** [comment]
Seems #8378 will fix the "Missing prebuilt package" error

File: recipes/gtk/all/conanfile.py
Comment (ericLemanissier) [inline recipes/gtk/all/conanfile.py:65]: Please also raise an exception in validate(self) if these two options are False

**Comment by SSE4** [comment]
we hit #8818 / #8667

**Comment by ericLemanissier** [comment]
@taoyouh can you please close the PR, wait 15 seconds, then reopen it, to retrigger the CI, now that https://github.com/conan-io/conan-center-index/pull/9051 is merged ?

**Comment by ericLemanissier** [comment]
blocked by https://github.com/conan-io/conan-center-index/pull/9124

**Comment by ericLemanissier** [comment]
And this is now unblocked.

**Comment by ericLemanissier** [comment]
I now get on windows 
```
FAILED: gtk/gtk_gtk-win32.rc.body_gtk-win32.rc.res 
"C:\Program Files (x86)\Windows Kits\10\bin\10.0.22000.0\x64\rc.EXE" "-IC:/.conan/913c1a/1/build_subfolder/gdk/win32/rc" "-IC:/.conan/913c1a/1/source_subfolder/gdk/win32/rc" "/nologo" "/fogtk/gtk_gtk-win32.rc.body_gtk-win32.rc.res" "gtk/gtk-win32.rc.body"
C:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\um\winnt.h(253) : error RC2188: C:\.conan\913c1a\1\build_subfolder\gtk\RCa05292(36) : fatal error RC1116: RC terminating after preprocessor errors
```
and gtk3 on linux:
```
/usr/bin/ld: /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/libcairo.a(cairo-pattern.c.o): in function `INT_cairo_pattern_create_for_surface':
cairo-pattern.c:(.text+0x1110): multiple definition of `cairo_pattern_create_for_surface'; /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/cairo/libcairo-sphinx.a(fdr.c.o):fdr.c:(.text+0x10a0): first defined here
/usr/bin/ld: /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/libcairo.a(cairo-pattern.c.o): in function `INT_cairo_pattern_destroy':
cairo-pattern.c:(.text+0x17b0): multiple definition of `cairo_pattern_destroy'; /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/cairo/libcairo-sphinx.a(fdr.c.o):fdr.c:(.text+0xb20): first defined here
/usr/bin/ld: /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/libcairo.a(cairo-pattern.c.o): in function `cairo_pattern_get_surface':
cairo-pattern.c:(.text+0x7b60): multiple definition of `cairo_pattern_get_surface'; /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/cairo/libcairo-sphinx.a(fdr.c.o):fdr.c:(.text+0x11c0): first defined here
/usr/bin/ld: /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/libcairo.a(cairo-surface.c.o): in function `cairo_surface_create_similar':
cairo-surface.c:(.text+0x2bd0): multiple definition of `cairo_surface_create_similar'; /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/cairo/libcairo-sphinx.a(fdr.c.o):fdr.c:(.text+0x1480): first defined here
/usr/bin/ld: /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/libcairo.a(cairo.c.o): in function `INT_cairo_create':
cairo.c:(.text+0x20): multiple definition of `cairo_create'; /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/cairo/libcairo-sphinx.a(fdr.c.o):fdr.c:(.text+0x290): first defined here
/usr/bin/ld: /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/libcairo.a(cairo.c.o): in function `INT_cairo_destroy':
cairo.c:(.text+0xf0): multiple definition of `cairo_destroy'; /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/cairo/libcairo-sphinx.a(fdr.c.o):fdr.c:(.text+0x900): first defined here
/usr/bin/ld: /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/libcairo.a(cairo.c.o): in function `INT_cairo_set_source_surface':
cairo.c:(.text+0x470): multiple definition of `cairo_set_source_surface'; /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/cairo/libcairo-sphinx.a(fdr.c.o):fdr.c:(.text+0x1300): first defined here
/usr/bin/ld: /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/libcairo.a(cairo.c.o): in function `INT_cairo_get_target':
cairo.c:(.text+0x26d0): multiple definition of `cairo_get_target'; /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/cairo/libcairo-sphinx.a(fdr.c.o):fdr.c:(.text+0xe80): first defined here
/usr/bin/ld: /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/libcairo.a(cairo.c.o): in function `cairo_get_group_target':
cairo.c:(.text+0x26f0): multiple definition of `cairo_get_group_target'; /root/.conan/data/cairo/1.17.4/_/_/package/5191ae0b0467f9b431b0c2c9b13702c5f2961ed2/lib/cairo/libcairo-sphinx.a(fdr.c.o):fdr.c:(.text+0xf90): first defined here
```

**Comment by SSE4** [comment]
```
"C:\J\w\BuildSingleReference@6/s\46a8e1\1\bin\glib-compile-resources.EXE" "gtk/gtk.gresources.xml" "--sourcedir" "C:/J/w/BuildSingleReference@6/s/4fb9b4/1/build_subfolder/gtk" "--sourcedir" "C:/J/w/BuildSingleReference@6/s/4fb9b4/1/source_subfolder/gtk" "--sourcedir" "gtk/theme/Default" "--sourcedir" "../source_subfolder/gtk" "--c-name" "_gtk" "--internal" "--generate" "--target" "gtk/gtkresources.c" "--manual-register" "--dependency-file" "gtk/gtkresources.c.d"
xml-stripblanks preprocessing requested, but XMLLINT is not set, and xmllint is not in PATH
```
something about `xmllint` package and environment variable XMLLINT?

File: recipes/gtk/all/conanfile.py
Comment (ericLemanissier) [inline recipes/gtk/all/conanfile.py:92]:
```suggestion
        self.build_requires("meson/0.60.2")
        self.build_requires("libxml2/2.9.12") # for xmllint
```
This should make the windows build go one step further

**Comment by ericLemanissier** [comment]
ok, I'm trying https://github.com/taoyouh/conan-center-index/pull/1

**Comment by ericLemanissier** [comment]
this is blocked by:
- https://github.com/conan-io/conan-center-index/pull/9335
- https://github.com/taoyouh/conan-center-index/pull/1

**Comment by ericLemanissier** [comment]
OK, so this is now unblocked. Can we give it another try?

**Comment by taoyouh** [comment]
My local try with `build=*` produces the following error on msys:

```
PS>  conan create . gtk/4.3.2@test/test --build=*
...
错误：无法准备事务处理 (无法满足依赖关系)
(Translated: "Error: Unable to prepare transaction (unable to satisfy dependency)")
:: 删除 pkgconf 破坏依赖 'pkgconf' （base-devel 需要）
(Translated: ":: remove pkgconf breaking dependency 'pkgconf' (required by base-devel)")
msys2/cci.latest:
msys2/cci.latest: ERROR: Package 'eee3fba89db6d777329de604625af8c30d46f080' build failed
msys2/cci.latest: WARN: Build folder C:\.conan\7ec140\1
ERROR: msys2/cci.latest: Error in build() method, line 122
        self._do_build()
while calling '_do_build', line 137
        self.run('bash -l -c "pacman -Rs $(pacman -Qsq %s) --noconfirm"' % package)
        ConanException: Error 1 while executing bash -l -c "pacman -Rs $(pacman -Qsq pkgconf) --noconfirm"
```

**Comment by SSE4** [comment]
@ericLemanissier any idea on how to fix these resource errors?

**Comment by ericLemanissier** [comment]
no, sorry !

**Comment by taoyouh** [comment]
Requires:
- #10304 to fix glib override
- #10322 to fix glib linking issue (undefined reference)
- Some update so that gdk-pixbuf uses glib 2.72.0 so that static-shared mismatch is fixed

File: recipes/gtk/all/patches/0001-fix-windows-rc-error.patch
Comment (ericLemanissier) [inline recipes/gtk/all/patches/0001-fix-windows-rc-error.patch:25]: So you managed to find the fix, great! 
Does it come from upstream ? did you submit it ?

Comment (taoyouh) [reply recipes/gtk/all/patches/0001-fix-windows-rc-error.patch:25]: No it does not come from upstream and I did not submit it yet. I think this is related to the version of Windows SDK, and I'm not very sure whether it is the correct solution. Will discuss this with upstream maintainers later.

I checked the log and find this line and noticed that I recently went into the same problem:
```
C:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\um\winnt.h(253) : error RC2188
```

Comment (taoyouh) [reply recipes/gtk/all/patches/0001-fix-windows-rc-error.patch:25]: I checked and they have already fixed this. Let me use their patch instead.

File: recipes/gtk/all/conanfile.py
Comment (ericLemanissier) [inline recipes/gtk/all/conanfile.py:87]:
```suggestion
        if tools.Version(self.version) >= "4.1.0":
            if not self.options.shared:
                raise ConanInvalidConfiguration("gtk supports only shared since 4.1.0")
```
cf https://github.com/GNOME/gtk/commit/14f0a0addb9a195bad2f8651f93b95450b186bd6

Comment (taoyouh) [reply recipes/gtk/all/conanfile.py:87]: Shall we also set the default option to shared?

File: recipes/gtk/all/conanfile.py
Comment (ericLemanissier) [inline recipes/gtk/all/conanfile.py:71]:
```suggestion
        if tools.Version(self.version) >= "4.1.0" or self.settings.os == "Windows":
            self.options.shared = True
```

**Comment by taoyouh** [comment]
> you could complete the system_libs attribute on windows, according to https://github.com/GNOME/gtk/blob/eaf4fb68f396d18fcb0934b0af1f77c483444c6b/gtk/meson.build#L1142-L1150 and
> 
> ```
> post_package(): WARN: [MISSING SYSTEM LIBS (KB-H043)] Library '.\bin\gtk-4-1.dll' links to system library 'setupapi' but it is not in cpp_info.system_libs.
> post_package(): WARN: [MISSING SYSTEM LIBS (KB-H043)] Library '.\bin\gtk-4-1.dll' links to system library 'imm32' but it is not in cpp_info.system_libs.
> post_package(): WARN: [MISSING SYSTEM LIBS (KB-H043)] Library '.\bin\gtk-4-1.dll' links to system library 'winmm' but it is not in cpp_info.system_libs.
> post_package(): WARN: [MISSING SYSTEM LIBS (KB-H043)] Library '.\bin\gtk-4-1.dll' links to system library 'crypt32' but it is not in cpp_info.system_libs.
> post_package(): WARN: [MISSING SYSTEM LIBS (KB-H043)] Library '.\bin\gtk-4-1.dll' links to system library 'comctl32' but it is not in cpp_info.system_libs.
> post_package(): WARN: [MISSING SYSTEM LIBS (KB-H043)] Library '.\bin\gtk-4-1.dll' links to system library 'dwmapi' but it is not in cpp_info.system_libs.
> ```

In my experience, Windows dynamic libraries will have all the static symbols inside it, which means, if DLL-A is built and linked to LIB-B, we don't need to link to LIB-B when we use DLL-A. I think when we use gtk-4-1.dll, we don't need to link to those libraries manually. Am I right?

**Review by ericLemanissier — DISMISSED** [review]
Great !

**Review by ericLemanissier — DISMISSED** [review]

**Review by ericLemanissier — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Pull request merged**
